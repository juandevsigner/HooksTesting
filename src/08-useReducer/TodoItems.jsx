import React from "react";

export const TodoItems = ({ todos, deleteTodo, onToggleTodo }) => {
  return (
    <>
      {todos.map(todo => (
        <li key={todo.id}>
          <span
            onClick={() => onToggleTodo(todo.id)}
            style={{ display: "block" }}
            aria-label="span"
          >
            {todo.done ? "âœ…" : null}
            {todo.description}
          </span>
          <button aria-label="delete" onClick={() => deleteTodo(todo.id)}>
            Delete Todo
          </button>
        </li>
      ))}
    </>
  );
};
