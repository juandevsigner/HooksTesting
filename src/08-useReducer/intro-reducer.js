

const initialState = [{
    id: 1,
    todo:"Aprender React",
    done: false
}]

const todoReducer =(state = initialState, action) =>{

return state
}

const newTodo ={
    id: 2,
    todo:"Aprender Redux",
    done: false
}

let todos = todoReducer()
todos = todoReducer(todos, )