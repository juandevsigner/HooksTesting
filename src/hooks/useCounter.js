import { useState } from "react"



export const useCounter =(initialValue = 10) =>{
    const [counter, setcounter] = useState(initialValue)

    const increment = (value) =>{
        setcounter(counter + value)
    }

    const decrement = () =>{
        if(counter <= 0) return
        setcounter(counter - 1)
    }

    const reset =() => {
        setcounter(initialValue)
    }
    return{
        increment,
        decrement,
        reset,
        counter
    }
}