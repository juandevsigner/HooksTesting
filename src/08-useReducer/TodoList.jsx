import React from "react";
import { TodoItems } from "./TodoItems";

export const TodoList = ({ todos, deleteTodo, onToggleTodo }) => {
  return (
    <ul>
      <TodoItems
        onToggleTodo={onToggleTodo}
        deleteTodo={deleteTodo}
        todos={todos}
      />
    </ul>
  );
};
