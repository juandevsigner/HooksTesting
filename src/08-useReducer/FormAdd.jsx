import { useState } from "react";
import { useForm } from "../hooks/useForm";

export const FormAdd = ({ handleTodo }) => {
  const { description, onInputChange, onHandleReset } = useForm({
    description: "",
  });

  const onSubmit = e => {
    e.preventDefault();

    if (description === "") return;

    const todo = {
      id: new Date().getTime() * 5,
      description,
      done: false,
    };

    handleTodo(todo);
    onHandleReset();
  };

  return (
    <form>
      <input
        onChange={onInputChange}
        value={description}
        name="description"
        type="text"
        placeholder="add todo here!"
      />
      <button onClick={onSubmit} type="submit">
        Add TODO{" "}
      </button>
    </form>
  );
};
