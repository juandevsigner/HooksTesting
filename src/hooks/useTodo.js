import { useReducer, useEffect } from "react";
import { todoReducer } from "../08-useReducer/todoReduce";

const init = () => {
  return JSON.parse(localStorage.getItem("todos")) || [];
};


export const useTodo =() =>{

const [todos, dispatch] = useReducer(todoReducer, [], init);
  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos) || []);
  }, [todos]);

const handleTodo = todo => {
    const action = {
      type: "Add Todo",
      payload: todo,
    };
    dispatch(action);
  };

  const deleteTodo = id => {
    const action = {
      type: "Delete Todo",
      payload: id,
    };
    dispatch(action);
  };

  const onToggleTodo = id => {
    const action = {
      type: "Toggle Todo",
      payload: id,
    };
    dispatch(action);
  };

  const todoCount = todos.length
  const todoPendingCount = todos.filter(todo => !todo.done).length

    return{
        todos, handleTodo, deleteTodo, onToggleTodo, todoCount, todoPendingCount
    }
}