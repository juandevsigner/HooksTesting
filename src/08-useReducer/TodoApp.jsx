import { useTodo } from "../hooks/useTodo";
import { FormAdd } from "./FormAdd";
import { TodoList } from "./TodoList";

export const TodoApp = () => {
  const {
    deleteTodo,
    handleTodo,
    onToggleTodo,
    todos,
    todoCount,
    todoPendingCount,
  } = useTodo();

  return (
    <>
      <h1>TodoApp {todoCount}</h1>
      <p>pending: {todoPendingCount}</p>
      <hr />
      <div>
        <FormAdd handleTodo={handleTodo} />
      </div>
      <div>
        <h4>All Todos</h4>
        <hr />
        <TodoList
          onToggleTodo={onToggleTodo}
          deleteTodo={deleteTodo}
          todos={todos}
        />
      </div>
    </>
  );
};
