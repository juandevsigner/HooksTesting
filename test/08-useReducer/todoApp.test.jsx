import { render, screen } from "@testing-library/react";
import { TodoApp } from "../../src/08-useReducer/TodoApp";
import { useTodo } from "../../src/hooks/useTodo";

jest.mock("../../src/hooks/useTodo");

describe("Todo App Test", () => {
  useTodo.mockReturnValue({
    deleteTodo: jest.fn(),
    handleTodo: jest.fn(),
    onToggleTodo: jest.fn(),
    todos: [
      { id: 1, description: "demo1", done: false },
      { id: 2, description: "demo2", done: false },
      { id: 3, description: "demo3", done: true },
    ],
    todoCount: 3,
    todoPendingCount: 2,
  });
  test("mostrar el componente correctamente", () => {
    const componenteApp = render(<TodoApp />);
    expect(componenteApp.container).toMatchSnapshot();
    expect(screen.getByText("demo1")).toBeTruthy();
    expect(screen.getByText("demo2")).toBeTruthy();
    expect(screen.getByText("âœ…demo3")).toBeTruthy();
  });
});
