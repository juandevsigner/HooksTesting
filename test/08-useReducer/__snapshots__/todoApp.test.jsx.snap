// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo App Test mostrar el componente correctamente 1`] = `
<div>
  <h1>
    TodoApp 
    3
  </h1>
  <p>
    pending: 
    2
  </p>
  <hr />
  <div>
    <form>
      <input
        name="description"
        placeholder="add todo here!"
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Add TODO
         
      </button>
    </form>
  </div>
  <div>
    <h4>
      All Todos
    </h4>
    <hr />
    <ul>
      <li>
        <span
          aria-label="span"
          style="display: block;"
        >
          demo1
        </span>
        <button
          aria-label="delete"
        >
          Delete Todo
        </button>
      </li>
      <li>
        <span
          aria-label="span"
          style="display: block;"
        >
          demo2
        </span>
        <button
          aria-label="delete"
        >
          Delete Todo
        </button>
      </li>
      <li>
        <span
          aria-label="span"
          style="display: block;"
        >
          âœ…
          demo3
        </span>
        <button
          aria-label="delete"
        >
          Delete Todo
        </button>
      </li>
    </ul>
  </div>
</div>
`;
